hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java

```diff
@@ -308,36 +371,91 @@ private void generateSnapshotDiffReport(final String jobId,
+        try {
+          addToObjectIdMap(fsKeyTable,
+                  tsKeyTable,
+                  Pair.of(false, deltaFilesForKeyOrFileTable),
+                  objectIdToKeyNameMapForFromSnapshot,
+                  objectIdToKeyNameMapForToSnapshot,
+                  objectIDsToCheckMap,
+                  tablePrefixes);
+        } catch (NativeLibraryNotLoadedException ex) {
+          //This code should be never executed.
+          throw new IllegalStateException(ex);
+        }
+      }
 
       if (bucketLayout.isFileSystemOptimized()) {
+        // add to object ID map for directory.
         final Table<String, OmDirectoryInfo> fsDirTable =
             fromSnapshot.getMetadataManager().getDirectoryTable();
         final Table<String, OmDirectoryInfo> tsDirTable =
             toSnapshot.getMetadataManager().getDirectoryTable();
+        tablesToLookUp = Collections.singletonList(fsDirTable.getName());
         final Set<String> deltaFilesForDirTable =
-            getDeltaFiles(fromSnapshot, toSnapshot,
-                Collections.singletonList(fsDirTable.getName()), fsInfo, tsInfo,
-                useFullDiff, tablePrefixes);
-        addToObjectIdMap(fsDirTable,
-            tsDirTable,
-            deltaFilesForDirTable,
-            objectIdToKeyNameMapForFromSnapshot,
-            objectIdToKeyNameMapForToSnapshot,
-            objectIDsToCheckMap,
-            tablePrefixes);
+            getDeltaFiles(fromSnapshot, toSnapshot, tablesToLookUp, fsInfo,
+                    tsInfo, useFullDiff, tablePrefixes);
+        if (!isNativeRocksToolsLoaded) {
+          deltaFilesForDirTable.addAll(getSSTFileListForSnapshot(
+                  fromSnapshot, tablesToLookUp));
+        }
+        try {
+          addToObjectIdMap(fsDirTable,
+              tsDirTable,
+              Pair.of(isNativeRocksToolsLoaded, deltaFilesForDirTable),
+              objectIdToKeyNameMapForFromSnapshot,
+              objectIdToKeyNameMapForToSnapshot,
+              objectIDsToCheckMap,
+              tablePrefixes);
+        } catch (NativeLibraryNotLoadedException e) {
+          try {
+            // Workaround to handle deletes if use of native rockstools for
+            // reading tombstone fails.
+            // TODO: [SNAPSHOT] Update Rocksdb SSTFileIterator to read to
+            //  read tombstone
+            deltaFilesForDirTable.addAll(getSSTFileListForSnapshot(
+                    fromSnapshot, tablesToLookUp));
+            addToObjectIdMap(fsDirTable,
+                    tsDirTable,
+                    Pair.of(false, deltaFilesForDirTable),
+                    objectIdToKeyNameMapForFromSnapshot,
+                    objectIdToKeyNameMapForToSnapshot,
+                    objectIDsToCheckMap,
+                    tablePrefixes);
+          } catch (NativeLibraryNotLoadedException ex) {
+            //This code should be never executed.
+            throw new IllegalStateException(ex);
+          }
+        }
```

Comment:
nit: This (line 427-458)  is duplicate of lines 384-415. You may move this to a function and call it from both the places.

Reply:
I believe we are doing redundant checks on the sst files, when bucket is FSOptimized. This would change a lot of things the way code has been implemented. Will take this up in the next PR.